fs/src/jsMain/kotlin/com/suparnatural/core/fs/FileSystem.kt

```diff
@@ -0,0 +1,325 @@
+                    stat.isFile() -> FileType.Regular
+                    stat.isDirectory() -> FileType.Directory
+                    else -> FileType.Unknown
+                }
+            }()
+        )
+    }
+
+    /**
+     * Returns stats for the resource at `pathComponent`.
+     */
+    actual fun stat(pathComponent: PathComponent): StatResult? = pathComponent.component?.let { stat(it) }
+
+    /**
+     *
+     * Returns the contents of the file located at `path`. The content is parsed according to `encoding`.
+     * For binary files, use `encoding` = [ContentEncoding.Base64].
+     *
+     */
+    actual fun readFile(path: String, encoding: ContentEncoding): String? =
+        if (encoding == ContentEncoding.Base64)
+        //nodejs base64 encoding on readFile doesn't decode it, it just encodes it again,
+        //so instead we have to decode it ourself
+            Buffer.from(readFileSync(fixPathString(path), encodingOptions(ContentEncoding.Ascii)) as String, "base64")
+                .toString("ascii")
+        else
+            readFileSync(fixPathString(path), encodingOptions(encoding)) as? String
+
+    /**
+     *
+     * Returns the contents of the file located at `pathComponent`. The content is parsed according to `encoding`.
+     * For binary files, use `encoding` = [ContentEncoding.Base64].
+     *
+     */
+    actual fun readFile(pathComponent: PathComponent, encoding: ContentEncoding): String? =
+        pathComponent.component?.let { readFile(it, encoding) }
+
+    /**
+     * Returns the contents of the file located at `path` as ByteArray.
+     */
+    actual fun readFile(path: String): ByteArray? =
+        readFileSync(fixPathString(path), encodingOptions(ContentEncoding.Utf8)).toString().encodeToByteArray()
+
+    /**
+     * Returns the contents of the file located at `pathComponent` as ByteArray.
+     */
+    actual fun readFile(pathComponent: PathComponent): ByteArray? = pathComponent.component?.let { readFile(it) }
+
+    /**
+     * tries to execute the given [callback]
+     * If [create] is true, then file is created if it does not exist.
+     * @return true if operation is successful, otherwise false.
+     */
+    private fun tryIfExists(path: String, create: Boolean, callback: () -> Unit): Boolean =
+        if (!create && !exists(fixPathString(path))) {
+            false
+        } else {
+            try {
+                callback()
+                true
+            } catch (error: Throwable) {
+                false
+            }
+        }
+
+    /**
+     * Writes `contents` to the file located at `path`. If `create` is true, then file is created if it does not exist.
+     * For binary files, use `encoding` = [ContentEncoding.Base64].
+     * * Returns true if operation is successful, otherwise false.
+     */
+    actual fun writeFile(path: String, contents: String, create: Boolean, encoding: ContentEncoding): Boolean =
+        tryIfExists(fixPathString(path), create) {
```

Comment:
We don't need to call `fixPathString` here since `tryIfExists` is already calling it. We can have a convention where the caller doesn't do any transformation and relies on the callee to do it or vice versa but not both ways.